---
output: 
  github_document:
    pandoc_args: --webtex
always_allow_html: yes
---

<!-- README.md is generated from README.Rmd. Please edit that file -->
[![Build status](https://ci.appveyor.com/api/projects/status/xwu8nket2u1pd5ij?svg=true)](https://ci.appveyor.com/project/hsmith9002/diffenrich)

[![Build Status](https://travis-ci.com/SabaLab/diffEnrich.svg?token=M2jELEzTYYqmxZMY6hpb&branch=master)](https://travis-ci.com/SabaLab/diffEnrich)


```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%",
  eval.after = 'fig.cap'
) 
```
# diffEnrich

## Introduction

The goal of diffEnrich is to compare functional enrichment between two experimentally-derived groups of genes or proteins. Given a list of gene symbols, *diffEnrich*  will
perform differential enrichment analysis using the Kyoto Encyclopedia of Genes
and Genomes (KEGG) REST API. This package provides a number of functions that are 
intended to be used in a pipeline (See Figure 1). Briefly, the user provides a KEGG formated species id for either human, mouse or rat, and the package will
download and clean species specific ENTREZ gene IDs and map them to their respective
KEGG pathways by accessing KEGG's REST API. KEGG's API is used ot guarantee the most up-to-date pathway data from KEGG. Next, the user will identify significantly
enriched pathways from two different gene sets, and finally, the user will identify 
pathways that are differentially enriched between the two gene sets. In addition to 
the analysis pipeline, this package also provides a plotting function. 

**Note on figure legends:** To view figure legends, hover mouse over image.

**The KEGG REST API**

KEGG is a database resource for understanding high-level functions 
of a biological system, such as a cell, an organism and an ecosystem, 
from genomic and molecular-level information [REF](https://www.kegg.jp/kegg/kegg1a.html). KEGG is an integrated database 
resource consisting of eighteen databases that are clustered into 4 main categories:
1) systems information (e.g. hierarchies and maps), 2) genomic information 
(e.g. genes and proteins), 3) chemical information (e.g. biochemical reactions), 
and 4) health information (e.g. human disease and drugs) [REF](https://www.kegg.jp/kegg/kegg1a.html). 

In 2012 KEGG released its first application programming interface (API), and has been
adding features and functionality ever since. There are benefits to using an API.
First, API's like KEGG's allow users to perform customized analyses with the most
up-to-date versions of the data contained in the database. In addition, accessing the
KEGG API is very easy using statistical programming tools like R or Python and
integrating data pulls into user's code makes the program reproducible. To 
further enforce reproducibilty *diffEnrich* adds a date and KEGG release tag to
all data files it generates from accessing the API. For update histories and release 
notes for the KEGG REST API please go [here](https://www.kegg.jp/kegg/rest/). 


```{r, echo=FALSE, fig.cap=cap, dpi=400}
suppressMessages(library(diagram))

## open plot
par(mar = c(0.5, 0.5, 0.5, 0.5))
openplotmat()
## define number of boxes at each level
elpos <- coordinates (c(3, 1, 1, 2, 2, 1, 1, 1))
## generate matrix of from-to arrow coordinates
fromto <- matrix(ncol = 2, byrow = TRUE,
                 data = c(1, 6,
                          2, 4,
                          3, 7,
                          4, 5,
                          5, 6,
                          5, 7,
                          6, 8,
                          7, 9,
                          8, 10, 
                          9, 10,
                          10, 11,
                          11, 12))
nr <- nrow(fromto)
arrpos <- matrix(ncol = 2, nrow = nr)

## Build flow chart
for (i in 1:nr){
  arrpos[i, ] <- straightarrow (to = elpos[fromto[i, 2], ],
                                from = elpos[fromto[i, 1], ],
                                lwd = 2, arr.pos = 0.6, arr.length = 0.5)
  textellipse(elpos[1,], 0.10, 0.04, lab = "geneList1", box.col = "purple",
              shadow.col = "purple4", shadow.size = 0.005, cex = 1.0)
  textrect(elpos[2,], 0.10, 0.03, lab = "get_kegg()", box.col = "steelblue1",
              shadow.col = "darkblue", shadow.size = 0.005, cex = 1.0)
  textellipse(elpos[3,], 0.10, 0.04, lab = "geneList2", box.col = "purple",
              shadow.col = "purple4", shadow.size = 0.005, cex = 1.0)
  textround(elpos[4,], 0.10, 0.03,lab = "KEGG REST API", box.col = "orange1",
            shadow.col = "orange4", shadow.size = 0.005, cex = 1.0)
  textellipse(elpos[5,], 0.15, 0.04, lab = c("cleaned gene-to-pathway", "map"), box.col = "firebrick1",
              shadow.col = "firebrick4", shadow.size = 0.005, cex = 1.0)
  textrect(elpos[6,], 0.10, 0.03,lab = "pathEnrich(List 1)", box.col = "steelblue1",
            shadow.col = "darkblue", shadow.size = 0.005, cex = 1.0)
  textrect(elpos[7,], 0.10, 0.03, lab = "pathEnrich(List 2)", box.col = "steelblue1",
           shadow.col = "darkblue", shadow.size = 0.005, cex = 1.0)
  textellipse(elpos[8,], 0.10, 0.04, lab = c("pathEnrich","result/object"), box.col = "firebrick1",
              shadow.col = "firebrick4", shadow.size = 0.005, cex = 1.0)
  textellipse(elpos[9,], 0.10, 0.04, lab = c("pathEnrich","result/object"),box.col = "firebrick1",
              shadow.col = "firebrick4", shadow.size = 0.005, cex = 1.0)
  textrect(elpos[10,], 0.10, 0.03, lab = "diffEnrich()", box.col = "steelblue1",
           shadow.col = "darkblue", shadow.size = 0.005, cex = 1.0)
  textellipse(elpos[11,], 0.10, 0.04, lab = c("diffEnrich","result/object"), box.col = "firebrick1",
              shadow.col = "firebrick4", shadow.size = 0.005, cex = 1.0)
  textrect(elpos[12,], 0.11, 0.03, lab = "plotFoldEnrichment", box.col = "steelblue1",
           shadow.col = "darkblue", shadow.size = 0.005, cex = 1.0)
  # text(0.15,0.05, "Figure 1. diffEnrich Analysis pipeline", cex = 0.75)
  text(0.05,0.87, "Step 1", cex = 1.2, font = 2)
  text(0.05,0.57, "Step 2", cex = 1.2, font = 2)
  text(0.05,0.35, "Step 3", cex = 1.2, font = 2)
  text(0.05,0.10, "Step 4", cex = 1.2, font = 2)
}

cap <- c("Figure 1. diffEnrich Analysis pipeline. Functions within the diffEnrich package are represented by blue rectangles. The data that must be provided by the user is represented by the pruple ovals. Data objects generated by a function in diffEnrich are representaed by red ovals. The external call of the get_kegg function to the KEGG REST API is represented in yellow.")
```


**Motivating experimental design for differential enrichment**

A recent [study](https://onlinelibrary.wiley.com/doi/full/10.1111/acer.13766) explored 
the hepatic mechanism for removal of toxic lipid aldehydes via conjugation with glutathione (Cheng et al., [2001](https://www.ncbi.nlm.nih.gov/pubmed/11488593?dopt=Abstract);
Gallagher et al., [2007](https://www.ncbi.nlm.nih.gov/pubmed/17553661?dopt=Abstract)) 
and the primary enzyme that catalyzes this conjugation, glutathione S‐transferase A4‐4 (GSTA4).
Specifically, the researchers examine the role of the GSTA4 gene on protein 
carbonylation and the progression of liver injury in a model consisting of long‐term 
(116 days) chronic ethanol (EtOH) consumption followed by a single EtOH binge.

Functional enrichment of carbonylated proteins for KEGG pathways was tested 
using a 1‐sided Fisher's exact test in each experimental group 
(2 genotypes × 3 treatment conditions = 6 experimental groups). Differential 
enrichment between experimental groups was examined using a 2‐sided Fisher's exact test.
For KEGG pathway enrichment, proteins were mapped to pathways using their UniProt ID, 
and databases were downloaded from the KEGG API.

Note: This example uses proteins as the feature of interest, however *diffEnrich*
currently only supports genes as the feature of interest.

Differential enrichment analysis of carbonylated proteins indicated that compared 
to chow‐fed animals, the most overrepresented group of proteins that was 
significantly adducted after long‐term EtOH exposure was ribosomal proteins.


## Installation

You can install the released version of diffEnrich from [CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("diffEnrich") 
```

## Example

### Step 1: Collect and clean pathways from KEGG API
First we will use the *get_kegg* function to access the KEGG REST API and download the data sets required to perform 
our downstream analysis. This function takes two arguments. The first, 'species' is required. Currently, *diffEnrich* supports three species, and the argument is a character string using the KEGG
code [REF](https://www.pnas.org/content/suppl/2008/09/11/0806162105.DCSupplemental/ST1_PDF.pdf): 
Homo sapiens (human), use 'hsa'; Mus musculus (mouse), use 'mmu', and Rattus norvegicus (rat), use 'rno'. The second,
'path' is also passed as a character string, and is the path of the directory in which the user would like to
write the data sets downloaded from the KEGG REST API. If the user does not provide a path, the data sets will 
be automatically written to the current working directory using the *here::here()* functionality. These data 
sets will be tab delimited files with a name describing the data, and for reproducibility, the date they were generated
and the version of KEGG when the API was accessed. In addition to these flat files, *get_kegg* will also
create a named list with the three relevant KEGG data sets. The names of this list will describe the data set.
For a detailed description of list elements use *?get_kegg*.


```{r, results='hide', echo=FALSE}
suppressMessages(library(diffEnrich))
```

```{r get_kegg_exp1}
## run get_kegg() using rat
kegg_rno <- get_kegg('rno')
```

Here are examples of the output files:
 
    kegg_to_pathway2019-04-26Release_90.0+_04-26_Apr_19.txt
    ncbi_to_kegg2019-04-26Release_90.0+_04-26_Apr_19.txt
    pathway_to_species2019-04-26Release_90.0+_04-26_Apr_19.txt

**Note:** Because it is assumed that the user might want to use the data sets generated by *get_kegg*, it is careful not to
overwrite data sets with exact names. *get_kegg* checks the path provided for data sets generated 'same-day/same-version', 
and if it finds even one of the three, it will not re-write any of the data sets. It will still however, let the user know
it is not writing out new data sets and still generate the named list object. Users can generate 'same-day/same-version'
data sets in different directories if they so choose.

```{r get_kegg_exp2}
## run get_kegg() using rat
kegg_rno <- get_kegg('rno')
```

```{r get_kegg_descriptions, echo=FALSE}
# Define column names
kegg_df_dat <- c("ncbi_to_kegg",
                 "kegg_to_pathway",
                 "pathway_to_species")

kegg_df_dscr <- c(paste0("ncbi gene ID",
                 " <-- mapped to --> ",
                 "KEGG gene ID"),
                 paste0("KEGG gene ID",
                 " <-- mapped to --> ",
                 "KEGG pathway ID"),
                 paste0("KEGG pathway ID",
                 " <-- mapped to --> ",
                 "KEGG pathway species description"))
# Make dataframe
kegg_df <- data.frame("get_kegg list object" = kegg_df_dat, 
                      "Object description" = kegg_df_dscr)
```

```{r get_kegg_kable, warning=FALSE, echo=FALSE, message=FALSE}
if (require(kableExtra)){
knitr::kable(kegg_df,
             caption = "Table 1. Description of data sets generated from accessing KEGG REST API", row.names = FALSE) %>%
  kableExtra::kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = T)
}
```

### Step 2: Perform enrichment analysis of individual gene sets

In this step we will use the *pathEnrich* function to identify KEGG pathways that are enriched (i.e. over-represented) based on a gene list of interest. User gene lists must also be character vectors and be formatted as ENTREZ gene IDs. The *clusterProfiler* package offers a nice function (*bitr*) that maps gene symbols and Ensembl IDs to ENTREZ gene IDs, and an example can be seen in their [vignette](https://yulab-smu.github.io/clusterProfiler-book/chapter5.html#supported-organisms).

```{r view_genelists}
## View sample gene lists from package data
head(geneLists$list1)
head(geneLists$list2)
```

This function may not always use the complete list of genes provided 
by the user. Specifically, it will only use the genes from the list provided that are also in
the most current species list pulled from the KEGG REST API using *get_kegg*, or from the older KEGG data loaded by the user from a 
previous *get_kegg* call. The *pathEnrich* function should be run at least twice, once for the genes of interest in list 1 and once for the genes of interest in list2. Each *pathEnrich* call generates a data frame summarizing the results of an enrichment analysis in which a Fisher's Exact test is used to identify which KEGG pathways are enriched within the user's list of interesting genes compared to all genes annotated to a KEGG pathway. P-values from the Fisher's Exact test are adjusted for multiple comparisons by controlling the False Discovery Rate (FDR) at 0.05.
 
```{r pathEnrich, warning=FALSE}
# run pathEnrich using kegg_rno
## List 1
list1_pe <- pathEnrich(gk_obj = kegg, gene_list = geneLists$list1)
## list2
list2_pe <- pathEnrich(gk_obj = kegg, gene_list = geneLists$list2) 
```

```{r pathEnrich_kable, warning=FALSE, echo=FALSE}
knitr::kable(head(list1_pe),
             caption = "Table 2. First 6 rows of list1_pe data frame generated using pathEnrich", row.names = FALSE) %>%
  kableExtra::kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = T)
```

*pathEnrich* generates a data frame with 9 columns described below. Details also provided in *pathEnrich* documentation. Use *?pathEnrich*.

```{r pathEnrich_description, echo=FALSE}
# get column names
names <- colnames(list1_pe)
# Define column names
description <- c("KEGG Pathway Identifier",
                 "Description of KEGG Pathway (provided by KEGG)",
                 "Number of Genes in KEGG Pathway",
                 "Number of Genes from gene list in KEGG Pathway",
                 "Number of Genes in KEGG Database",
                 "Number of Genes from gene list in KEGG Database",
                 "Expected number of genes from list to be in KEGG pathway by chance",
                 "P-value for enrichment within the KEGG pathway for list genes",
                 "False Discovery Rate (Benjamini and Hochberg) to account for multiple testing across KEGG pathways",
                 "Ratio of number of genes observed from the gene list annotated to the KEGG pathay to the number of genes expected from the gene list annotated to the KEGG pathway if ther was no enrichment (i.e. KEGG_PATHWAY_in_list/expected)")
# Make dataframe
df1 <- data.frame(names, description)
colnames(df1) <- c("Column Names", "Column Description") 
```

```{r pathEnrich_vars_kable, warning=FALSE, echo=FALSE}
knitr::kable(df1,
             caption = "Table 3. Description of columns is dataframe generated by pathEnrich", row.names = FALSE) %>%
  kableExtra::kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = F)
```

### Step 3: Identify differentially enriched KEGG pathways

The *diffEnrich* function will merge two results from the *pathEnrich* calls generated above. Specifically, the data frame
'list1_pe' and the data frame 'list2_pe' will be merged by the following columns:
"KEGG_PATHWAY_ID", "KEGG_PATHWAY_description", "KEGG_PATHWAY_cnt", "KEGG_DATABASE_cnt". This merged data set will then be used to perform differential enrichment using the same method and p-value calculation as described above. Users do have the option of choosing a method 
for multiple testing adjustment. Users can choose from those supported by *stats::p.adjust* for multiple correction options, and the default is the False Discovery Rate (Benjamini and Hochberg, [1995](http://www.jstor.org/stable/2346101)).

```{r diffEnrich}
## Perform differential enrichment
diff_enrich <- diffEnrich(list1_pe = list1_pe, list2_pe = list2_pe, method = 'none') 
```

```{r diffEnrich_kable, warning=FALSE, echo=FALSE}
knitr::kable(head(diff_enrich),
             caption = "Table 4.First 6 rows from data frame generated by diffEnrich", row.names = FALSE) %>%
  kableExtra::kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = F)
```


```{r diffEnrich_descriptions, echo=FALSE}
## get vector of column names
de_col_names <- names(diff_enrich)
de_col_descr <- c("KEGG Pathway Identifier",
                  "Description of KEGG Pathway (provided by KEGG)",
                  "Number of Genes in KEGG Pathway",
                  "Number of Genes in KEGG Database",
                  "Number of Genes from gene list 1 in KEGG Pathway",
                  "Number of Genes from gene list 1 in KEGG Database",
                  "Expected number of genes from list 1 to be in KEGG pathway by chance",
                  "P-value for enrichment of list 1 genes related to KEGG pathway",
                  "False Discovery Rate (Benjamini and Hochberg) of enrich_p_list1 to account for multiple testing across KEGG pathways",
                  "Ratio of number of genes observed from the gene list 1 annotated to the KEGG pathay to the number of genes expected from the gene list 1 annotated to the KEGG pathway if ther was no enrichment (i.e. KEGG_PATHWAY_in_list1/expected_list1)",
                  "Number of Genes from gene list 2 in KEGG Pathway",
                  "Number of Genes from gene list 2 in KEGG Database",
                  "Expected number of genes from list 2 to be in KEGG pathway by chance",
                  "P-value for enrichment of list 2 genes related to KEGG pathway",
                  "False Discovery Rate (Benjamini and Hochberg) of enrich_p_list2 to account for multiple testing across KEGG pathways",
                  "Ratio of number of genes observed from the gene list 2 annotated to the KEGG pathay to the number of genes expected from the gene list 2 annotated to the KEGG pathway if ther was no enrichment (i.e. KEGG_PATHWAY_in_list2/expected_list2)",
                  "Odds of a gene from list 2 being from this KEGG pathway / Odds of a gene from list 1 being from this KEGG pathway",
                  "P-value for differential enrichment of this KEGG pathway between list 1 and list 2",
                  "Multiple testing adjustment of diff_enrich_p (default = False Discovery Rate (Benjamini and Hochberg))")
# Make dataframe
de_descr_df <- data.frame(de_col_names, de_col_descr)
colnames(de_descr_df) <- c("Column Names", "Column Description") 
```

```{r diffEnrich_vars_kable, warning=FALSE, echo=FALSE}
knitr::kable(de_descr_df,
             caption = "Table 5. Description of columns is dataframe generated by diffEnrich", row.names = FALSE) %>%
  kableExtra::kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = F)
```

The result of the *diffEnrich* call is a data frame with the estimated odds ratio generated by the Fisher's Exact test 
and the associated p-value. 

### Step 4: Plot fold enrichment

*plotFoldEnrichment* generates a grouped bar plot using ggplot2 and the
*ggnewscale* package. There are 3 arguments: 1) *de_res* is the dataframe generated from the *diffEnrich* function, 2) *pval* is the threshold for the adjusted p-value associated with differential enrichment that will filter which KEGG pathways to plot, and 3) after filtering based on *pval* *N* tells the function how many pathways to plot. It is important to make a note that the significance of the fold change is associated with the number of genes in the gene list. Notice that in this example the pathways in gene list 2 have smaller fold changes (shorter bars) than those in list 1, but that many of them are more significant (darker blue). This is because there are more genes in gene list 2 compred to gene list 1.  

```{r echo=FALSE}
cap_plot <- c("Figure 2. Fold enrichment stratified by gene list. KEGG pathways are plotted on the y-axis and fold
enrichment is plotted on the x-axis. Each KEGG pathway has a bar depicting
its fold enrichment in list 1 (red) and its fold enrichment in list 2 (blue).
The transparency of the bars correspond to the unadjusted p-value for the
pathway's enrichment in the given list. The p-value presented as text to the
right of each pair of bars is the adjusted p-value (user defined: default is FDR) associated with the
differential enrichment of the pathway between the two lists, and the pathways
are ordered from top to bottom by this p-value (i.e. smallest p-value on top
of plot, and largest p-value on bottom of plot). The dotted line represents a fold enrichment of 1. Finally, the number of genes used
for analysis from each gene list (recall that this number may not be the same as the number of
genes in the user's original list) are reported below their respective p-values
in the legend.")
```

```{r plotFoldEnrichment, dpi=400, message=FALSE, fig.cap=cap_plot}
## Plot fold enrichment
plotFoldEnrichment(de_res = diff_enrich, pval = 0.05, N = 5)
```

