
<!-- README.md is generated from README.Rmd. Please edit that file -->
diffEnrich
==========

The goal of diffEnrich is simple. Given a list of gene symbols, *diffEnrich* will perform differential enrichment ananlysis using the Kyoto Encyclopedia of Genes and Genomes (KEGG) REST API. This package provides a number of functions that are intended to be used in a pipeline (See Figure N). Breifly, the workflow will download and clean species specific ENTREZ gene IDs and map them to their respective KEGG pathways by accessing KEGG's REST API. This way the user will always have the most up-to-date pathway data from KEGG. Next, the user will identify significantly enriched pathways from two different gene sets, and finally, the user will identify pathways that are differentially expressed between the two gene sets. In addition to the analysis pipeline, this package also provides functions for data visualizations.

development of methods and software for differential enrichment analysis

Papers from Saba Lab where differentially enrichment has been used: 1. [Shearn et al 2018 - ACER](https://onlinelibrary.wiley.com/doi/full/10.1111/acer.13766)

Installation
------------

You can install the released version of diffEnrich from [CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("diffEnrich")
```

Example
-------

First we will use the *get\_kegg* function access the KEGG REST API and download the data sets required to perform our downstream analysis. This function takes two arguments. The first, 'species' is required and is the species of interest. Currently, *diffEnrich* supports three species, and the argument is a character string using the KEGG code [REF](https://www.pnas.org/content/suppl/2008/09/11/0806162105.DCSupplemental/ST1_PDF.pdf): Homo sapiens (human), use 'hsa'; Mus musculus (mouse), use 'mmu', and Rattus norvegicus (rat), use 'rno'. The second, 'path' is also passed as a character string, and is the path of the directory in which the user would like to write the data sets downloaded from the KEGG REST API. If the user does not provide a path, the data sets will be automatically written to the current working directory using the *here::here()* functionality. These data sets will be tab delimited files with a name describing the data, and for reproducibility, the date they were generated and the version of KEGG when the API was accessed. In addition to these flat files, *get\_kegg* will also create a named list with the three relavent KEGG data sets. The names of this list will describe the the data set.

``` r
suppressMessages(library(diffEnrich))

## run get_kegg() using rat
kegg_rno <- get_kegg('rno')
#> 3 data sets will be written as tab delimited text files
#> File location: /Users/smiharry/Documents/packages/diffEnrich
#> Kegg Release: Release_90.0+_06-24_Jun_19
```

Here are the files:

    kegg_to_pathway2019-04-26Release_90.0+_04-26_Apr_19.txt
    ncbi_to_kegg2019-04-26Release_90.0+_04-26_Apr_19.txt
    pathway_to_species2019-04-26Release_90.0+_04-26_Apr_19.txt

**Note:** Because it is assumed that the user might want to use the data sets generated by *get\_kegg*, it is careful not to overwrite data sets with exact names. *get\_kegg* checks the path provided for data sets generated 'same-day/same-version', and if it finds even one of the three, it will not re-write any of the data sets. It will still however, let the user know it is not writing out new data sets and still generate the named list object. Users can of course generate 'same-day/same-version' data sets in different directories if they so choose.

``` r
## run get_kegg() using rat
kegg_rno <- get_kegg('rno')
#> These files already exist in your working directory. New files will not be generated.
#> Kegg Release: Release_90.0+_06-24_Jun_19
```
