---
title: "diffEnrich by example"
author: "Harry Smith"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{diffEnrich by example}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  %\VignetteDepends{kableExtra}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
options(knitr.table.format = "html") 
```

# diffEnrich

The goal of diffEnrich is simple. Given a list of gene symbols, *diffEnrich*  will
perform differential enrichment analysis using the Kyoto Encyclopedia of Genes
and Genomes (KEGG) REST API. This package provides a number of functions that are 
intended to be used in a pipeline (See Figure N). Briefly, the workflow will
download and clean species specific ENTREZ gene IDs and map them to their respective
KEGG pathways by accessing KEGG's REST API. This way the user will always have 
the most up-to-date pathway data from KEGG. Next, the user will identify significantly
enriched pathways from two different gene sets, and finally, the user will identify 
pathways that are differentially expressed between the two gene sets. In addition to 
the analysis pipeline, this package also provides functions for data visualizations.


development of methods and software for differential enrichment analysis

Papers from Saba Lab where differentially enrichment has been used:
1. [Shearn et al 2018 - ACER](https://onlinelibrary.wiley.com/doi/full/10.1111/acer.13766)

## Installation

You can install the released version of diffEnrich from [CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("diffEnrich") 
```

## Example

### Step 1: Collect and clean pathways from KEGG API
First we will use the *get_kegg* function access the KEGG REST API and download the data sets required to perform 
our downstream analysis. This function takes two arguments. The first, 'species' is required and is the species of 
interest. Currently, *diffEnrich* supports three species, and the argument is a character string using the KEGG
code [REF](https://www.pnas.org/content/suppl/2008/09/11/0806162105.DCSupplemental/ST1_PDF.pdf): 
Homo sapiens (human), use 'hsa'; Mus musculus (mouse), use 'mmu', and Rattus norvegicus (rat), use 'rno'. The second,
'path' is also passed as a character string, and is the path of the directory in which the user would like to
write the data sets downloaded from the KEGG REST API. If the user does not provide a path, the data sets will 
be automatically written to the current working directory using the *here::here()* functionality. These data 
sets will be tab delimited files with a name describing the data, and for reproducibility, the date they were generated
and the version of KEGG when the API was accessed. In addition to these flat files, *get_kegg* will also
create a named list with the three relevant KEGG data sets. The names of this list will describe the the data set.


```{r get_kegg_exp1}
suppressMessages(library(diffEnrich))

## run get_kegg() using rat
kegg_rno <- get_kegg('rno')
```

Here are the files:
 
    kegg_to_pathway2019-04-26Release_90.0+_04-26_Apr_19.txt
    ncbi_to_kegg2019-04-26Release_90.0+_04-26_Apr_19.txt
    pathway_to_species2019-04-26Release_90.0+_04-26_Apr_19.txt

**Note:** Because it is assumed that the user might want to use the data sets generated by *get_kegg*, it is careful not to
overwrite data sets with exact names. *get_kegg* checks the path provided for data sets generated 'same-day/same-version', 
and if it finds even one of the three, it will not re-write any of the data sets. It will still however, let the user know
it is not writing out new data sets and still generate the named list object. Users can of course generate 'same-day/same-version'
data sets in different directories if they so choose.

```{r get_kegg_exp2}
## run get_kegg() using rat
kegg_rno <- get_kegg('rno')
```

### Step 2: Perform individual enrichment analysis

In this step we will use the *pathEnrich* function to identify KEGG pathways that are enriched (i.e. over-represented) based on a list of genes we
are interested in and based on a list of background genes. This function may not always use the complete list of genes provided 
by the user. Specifically, it will only use the genes from the list provided that are also in
the most current species list pulled from the KEGG REST API using *get_kegg*, or from the older KEGG data loaded by the user from a 
previous *get_kegg* call. The *pathEnrich* function should be run at least twice, once for the genes of interest and once for the background.
Each *pathEnrich* call generates a data frame summarizing the results of an enrichment analysis in which a Fisher's
Exact test is used to identify which KEGG pathways are enriched by the user's list of interesting genes with respect to background
enrichment, and a p-value is calculated using a hypergeometric distribution (Formula 1). P-values are adjusted for multiple comparisons by controlling the False Discovery Rate (FDR) at 0.05.

Formula 1:

$$ P = 1-\sum_{i=0}^{k-1} \frac{ \binom{m} {i} \binom{N-M} {n-i} } {\binom{N} {n} } $$

where N is the total number of genes in the background, M is the number of genes within that background that also in the gene set of interest, n is the size of the list of genes of interest and k is the number of genes within that are in the background. The background by default is all the genes that have annotation. 
 
```{r pathEnrich, warning=FALSE}
# run pathEnrich using kegg_rno
## List 1
list1_pe <- pathEnrich(gk_obj = kegg, gene_list = geneLists$list1)
## list2
list2_pe <- pathEnrich(gk_obj = kegg, gene_list = geneLists$list2)
```

```{r pathEnrich_kable, warning=FALSE, echo=FALSE}
if (require(kableExtra)){
knitr::kable(head(list1_pe),
             caption = "Table 1. Head of list1_pe data frame generated using pathEnrich", row.names = FALSE) %>%
  kableExtra::kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = T)
}
```

*pathEnrich* generates a data frame with 9 columns described below.

```{r pathEnrich_description, echo=FALSE}
# get column names
names <- colnames(list1_pe)
# Define column names
description <- c("KEGG Pathway Identifier",
                 "Description of KEGG Pathway (provided by KEGG)",
                 "Number of Genes in KEGG Pathway",
                 "Number of Genes from gene list in KEGG Pathway",
                 "Number of Genes in KEGG Database",
                 "Number of Genes from gene list in KEGG Database",
                 "Expected number of genes from list to be in KEGG pathway by chance (i.e., not enriched)",
                 "P-value for enrichment of list genes related to KEGG pathway",
                 "False Discovery Rate (Benjamini and Hochberg) to account for multiple testing across KEGG pathways",
                 "KEGG_PATHWAY_in_list/expected")
# Make dataframe
df1 <- data.frame(names, description)
colnames(df1) <- c("Column Names", "Column Description")
```

```{r pathEnrich_vars_kable, warning=FALSE, echo=FALSE}
if (require(kableExtra)){
knitr::kable(df1,
             caption = "Table 2. Description of columns is dataframe generated by pathEnrich", row.names = FALSE) %>%
  kableExtra::kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = F)
}
```

### Step 3: Identify differentially enriched KEGG pathways

The diffEnrich function will merge the results from the *pathEnrich* calls generated above. Specifically, the data frame
'list1_pe' and the data frame 'list2_pe' will be merged using the base *merge* functionality and be merged by the following columns:
'c("KEGG_PATHWAY_ID", "KEGG_PATHWAY_description", "KEGG_PATHWAY_cnt", "KEGG_DATABASE_cnt")'. This merged data set will then be used to perform differential enrichment using the same method and p-value calculation as described above. Users do have the option of choosing a method 
for multiple testing adjustment. 

```{r diffEnrich}
## Perform differential enrichment
diff_enrich <- diffEnrich(list1_pe = list1_pe, list2_pe = list2_pe, method = 'none')
```

```{r diffEnrich_kable, warning=FALSE, echo=FALSE}
if (require(kableExtra)){
knitr::kable(head(diff_enrich),
             caption = "Table 3. Description of columns is data frame generated by diffEnrich", row.names = FALSE) %>%
  kableExtra::kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = F)
}
```

The result of the *diffEnrich* call is the merged data frame with the estimated odds ration generated by the Fisher's Exact test 
and the associated p-value. Users can choose from those supported by *stats::p.adjust*, and the defualt is the False Discovery Rate (Benjamini and Hochberg).
## Figures

### Step 4: Plot fold enrichment

```{r plotFoldEnrichment, message=FALSE, warning=FALSE, fig.height=6, fig.width=8}
## Plot fold enrichment
plotFoldEnrichment(de_res = diff_enrich, pval = 0.05, N = 5)
```


You can enable figure captions by `fig_caption: yes` in YAML:

    output:
      rmarkdown::html_vignette:
        fig_caption: yes

Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**.


